File: src/union_find_set/union_find_set.py
from typing import Iterable, Any, Dict


class Node:
    def __init__(self, val: Any):
        self.val = val


class UnionFindSet:
    def __init__(self, iterable: Iterable[Any]):
        self.nodes: Dict[Any, Node] = {}
        self.parent_map: Dict[Node, Node] = {}
        self.size_map: Dict[Node, int] = {}

        for item in iterable:
            node = Node(item)
            self.nodes[item] = node
            self.parent_map[node] = node
            self.size_map[node] = 1

    def find_head(self, node):
        """返回给定节点所在集合的头节点
        在找到头节点后, 会将node-head的链条扁平化
        """
        stack = []
        # 寻找node的头节点
        while node is not self.parent_map[node]:
            stack.append(node)
            node = self.parent_map[node]
        # 扁平化node-head的链条
        while stack:
            self.parent_map[stack.pop()] = node
        return node

    def same(self, a, b):
        """判断两个值是否在同一个集合中"""
        return self.find_head(self.nodes[a]) == self.find_head(self.nodes[b])

    def union(self, a, b):
        """合并a, b所在的集合"""
        ahead = self.find_head(self.nodes[a])
        bhead = self.find_head(self.nodes[b])

        # a, b不在同一个集合中才需要合并
        if ahead is not bhead:
            asize = self.size_map[ahead]
            bsize = self.size_map[bhead]

            largest_head, smaller_head = (ahead, bhead) if asize > bsize else (bhead, ahead)

            # 将更小的集合连接到更大的集合头, 避免出现长链表
            self.parent_map[smaller_head] = largest_head
            self.size_map[largest_head] = asize + bsize
            # smallerHead 不在是头节点, 删除器size信息
            self.size_map.pop(smaller_head)

    def __len__(self):
        return len(self.size_map)

    def __bool__(self):
        return len(self.size_map) > 0

File: src/union_find_set/UnionFindSet.java
package union_find_set;

import java.util.Collection;
import java.util.HashMap;
import java.util.Stack;

/**
 * @author liuchongwei
 * @email lcwliuchongwei@qq.com
 * @date 2022-01-27
 */
public class UnionFindSet<T> {
    static class Node<T> {
        T value;

        public Node(T value) {
            this.value = value;
        }
    }

    // 记录值对应的节点
    private HashMap<T, Node<T>> nodes;

    // 记录父节点映射
    private HashMap<Node<T>, Node<T>> parentMap;

    // 记录头节点大小
    private HashMap<Node<T>, Integer> sizeMap;

    public UnionFindSet(Collection<T> collection) {
        nodes = new HashMap<>();
        parentMap = new HashMap<>();
        sizeMap = new HashMap<>();

        for (T value : collection) {
            final Node<T> node = new Node<>(value);
            nodes.put(value, node);
            parentMap.put(node, node);
            sizeMap.put(node, 1);
        }
    }

    /**
     * 返回给定节点所在集合的头节点
     * <p>
     * 在找到头节点后, 会将node-head的链条扁平化
     */
    public Node<T> findHead(Node<T> node) {
        Stack<Node<T>> stack = new Stack<>();
        // 寻找node的头节点
        while (node != parentMap.get(node)) {
            stack.push(node);
            node = parentMap.get(node);
        }
        // 扁平化node-head的链条
        while (!stack.empty()) {
            parentMap.put(stack.pop(), node);
        }
        return node;
    }

    /**
     * 判断两个值是否在同一个集合中
     */
    public boolean isSameSet(T a, T b) {
        return findHead(nodes.get(a)) == findHead(nodes.get(b));
    }

    /**
     * 合并a, b所在的集合
     */
    public void union(T a, T b) {
        final Node<T> aHead = findHead(nodes.get(a));
        final Node<T> bHead = findHead(nodes.get(b));

        // a, b不在同一个集合中才需要合并
        if (aHead != bHead) {
            final Integer aSize = sizeMap.get(aHead);
            final Integer bSize = sizeMap.get(bHead);

            Node<T> largestHead = aSize > bSize ? aHead : bHead;
            Node<T> smallerHead = largestHead == aHead ? bHead : aHead;

            // 将更小的集合连接到更大的集合头, 避免出现长链表
            parentMap.put(smallerHead, largestHead);
            sizeMap.put(largestHead, aSize + bSize);
            // smallerHead 不在是头节点, 删除器size信息
            sizeMap.remove(smallerHead);
        }
    }

    /**
     * 返回并查集中有多少个集合
     */
    public int size() {
        return sizeMap.size();
    }
}